src/Agent.Worker/StepsRunner.cs

```diff
@@ -76,25 +77,46 @@ public sealed class StepsRunner : AgentService, IStepsRunner
                 step.ExecutionContext.Variables.RecalculateExpanded(out expansionWarnings);
                 expansionWarnings?.ForEach(x => step.ExecutionContext.Warning(x));
                 List<OperationCanceledException> allCancelExceptions = new List<OperationCanceledException>();
-                try
+                var timeout = step.TimeoutInMinutes;
```

Comment:
Thank you, this PR is getting close to the way I would like to see it. Now that I see it in this revision, I think there is a further significant refinement that should be made.

In JobRunner.cs can you pass the timeout to the factory method for the execution context instead:
```
jobContext.CreateChild(
    taskInstance.InstanceId,
    taskInstance.DisplayName,
    taskInstance.TimeoutInMinutes > 0 ? TimeSpan.FromMinutes(taskInstance.TimeoutInMinutes) : Timeout.InfiniteTimeSpan)
```

That will enable a few additional cleanups:
 * Steps runner will not be responsible for creating/linking the token.
 * The timeout token can be created/linked implicitly when `step.ExecutionContext.Start()` is called. The logic to manage the tokens should live inside the execution context class.
 * The changes to BuildJobExtension, JobExtensionRunner, ReleaseJobExtension, TaskRunner, and IStep can all be reverted.

ExecutionContext should expose one additional property `bool TimeoutExceeded { get; }`. Internally the property can check the private timeout token.

It doesn't look like the actual timeout or timeout token needs to be exposed publicly on any classes at all.

The impact to steps runner then becomes much simpler:
```
try
{
    await step.RunAsync();
}
catch (OperationCanceledException ex) when (step.ExecutionContext.TimeoutExceeded)
{
    [...]
}
catch (Exception ex)
{
    [...]
}
```

...leveraging the exception inspection language feature (`when ([...])`) will clean up the exception handling logic a little as well.

@TingluoHuang and @patrickcarnahan, thoughts?

@vithati pending confirmation from @TingluoHuang and/or @patrickcarnahan, it would be good to add a unit test for StepsRunner.